### PR TITLE
[Dataset] disable pandas as default block type

### LINKED ISSUE #22177: [Bug] datasets_preprocess_ingest failed in master due to pyarrow==7.0.0
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

https://buildkite.com/ray-project/periodic-ci/builds/2676#65781171-8c7f-4d05-a638-018f400047d0

### Versions / Dependencies

master

### Reproduction script

core nightly

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

The hypothesis is that pandas' deepcopy doesn't work as expected. specifically it still returns the full base copy even if the dataframe only has a a sliced view. this caused the shuffle-reduce to fail in our nightly test.

TODO: verify this fixed the object size too large issue.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
Lgtm if the test passes

**Comment by ericl** [comment]
It seems the flag might not completely disable pandas blocks. You can try:

```
 @DeveloperAPI
diff --git a/python/ray/data/dataset.py b/python/ray/data/dataset.py
index c997dbe862..f9f7c9c789 100644
--- a/python/ray/data/dataset.py
+++ b/python/ray/data/dataset.py
@@ -253,9 +253,10 @@ class Dataset(Generic[T]):
                 # bug where we include the entire base view on serialization.
                 view = block.slice(start, end, copy=batch_size is not None)
                 if batch_format == "native":
-                    # Always promote Arrow blocks to pandas for consistency.
-                    if isinstance(view, pa.Table) or isinstance(view, bytes):
-                        view = BlockAccessor.for_block(view).to_pandas()
+                    if context.enable_pandas_block:
+                        # Always promote Arrow blocks to pandas for consistency.
+                        if isinstance(view, pa.Table) or isinstance(view, bytes):
+                            view = BlockAccessor.for_block(view).to_pandas()
                 elif batch_format == "pandas":
                     view = BlockAccessor.for_block(view).to_pandas()
                 elif batch_format == "pyarrow":
@@ -1894,14 +1895,16 @@ class Dataset(Generic[T]):
         import pyarrow as pa
 
         time_start = time.perf_counter()
+        context = DatasetContext.get_current()
 
         def format_batch(batch: Block, format: str) -> BatchType:
             if batch_format == "native":
-                # Always promote Arrow blocks to pandas for consistency, since
-                # we lazily convert pandas->Arrow internally for efficiency.
-                if isinstance(batch, pa.Table) or isinstance(batch, bytes):
-                    batch = BlockAccessor.for_block(batch)
-                    batch = batch.to_pandas()
+                if context.enable_pandas_block:
+                    # Always promote Arrow blocks to pandas for consistency, since
+                    # we lazily convert pandas->Arrow internally for efficiency.
+                    if isinstance(batch, pa.Table) or isinstance(batch, bytes):
+                        batch = BlockAccessor.for_block(batch)
+                        batch = batch.to_pandas()
                 return batch
             elif batch_format == "pandas":
                 batch = BlockAccessor.for_block(batch)
```

Which should improve this, but maybe double check by adding some asserts into the pandas block to make sure it's never used from that test.

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -1897,14 +1898,15 @@ def iter_batches(
...
                 return batch
             elif batch_format == "pandas":
                 batch = BlockAccessor.for_block(batch)
+                if context.enable_pandas_block:
+                    # Always promote Arrow blocks to pandas for consistency, since
+                    # we lazily convert pandas->Arrow internally for efficiency.
+                    if isinstance(batch, pa.Table) or isinstance(batch, bytes):
+                        batch = BlockAccessor.for_block(batch)
+                        batch = batch.to_pandas()
                 return batch.to_pandas()
```
Comment (ericl) [inline python/ray/data/dataset.py:1910]: Does this do anything, given we already call to_pandas below?

**Comment by scv119** [comment]
hmm the build infra generated wrong wheels for this pr, specifically, the wheels 
https://ray-ci-artifact-pr-public.s3.amazonaws.com/35bd0e671c369776913db6617356002351a9fe37/tmp/artifacts/.whl/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
has commit b0452c6189b617587df95a6aaa6246509c9f3eb0

**Comment by jjyao** [comment]
> hmm the build infra generated wrong wheels for this pr, specifically, the wheels https://ray-ci-artifact-pr-public.s3.amazonaws.com/35bd0e671c369776913db6617356002351a9fe37/tmp/artifacts/.whl/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl has commit [b0452c6](https://github.com/ray-project/ray/commit/b0452c6189b617587df95a6aaa6246509c9f3eb0)

Yea, it still happens occasionally and we have an issue for it.

**Pull request closed without merging**